spec/services/calculation_orchestrator_spec.rb

```diff
@@ -4,58 +4,112 @@
+        per_participant: 323.17,
+      },
+      {
+        band: "B",
+        participants: 0,
+        monthly: 0.0,
+        per_participant: 391.60,
+      },
+      {
+        band: "C",
+        participants: 0,
+        monthly: 0.0,
+        per_participant: 386.40,
+      },
+    ]
+  end
+  let(:output_payments) do
+    [
+      {
+        band: "A",
+        participants: 10,
+        per_participant: 119.40,
+        subtotal: 1194.0,
+      },
+      {
+        band: "B",
+        participants: 0,
+        per_participant: 117.48,
+        subtotal: 0.0,
+      },
+      {
+        band: "C",
+        participants: 0,
+        per_participant: 115.92,
+        subtotal: 0.0,
+      },
+    ]
+  end
+  let(:other_fees) do
     {
-      service_fees: [
-        {
-          service_fee_monthly: 22_288.0,
-          service_fee_per_participant: 323.0,
-          service_fee_total: 646_349.0,
-        },
-        {
-          service_fee_monthly: 0.0,
-          service_fee_per_participant: 392.0,
-          service_fee_total: 0.0,
-        },
-        {
-          service_fee_monthly: 0.0,
-          service_fee_per_participant: 386.0,
-          service_fee_total: 0.0,
-        },
-      ],
-      output_payments: [
-        {
-          per_participant: 597.0,
-          started: {
-            retained_participants: 10,
-            per_participant: 119.0,
-            subtotal: 1194.0,
-          },
-        },
-        {
-          per_participant: 587.0,
-          started: {
-            retained_participants: 0,
-            per_participant: 117.0,
-            subtotal: 0.0,
-          },
-        },
-        {
-          per_participant: 580.0,
-          started: {
-            retained_participants: 0,
-            per_participant: 116.0,
-            subtotal: 0.0,
-          },
-        },
-      ],
       uplift: {
+        participants: 10,
         per_participant: 100.0,
-        sub_total: 1000.0,
+        subtotal: 1000.0,
       },
     }
   end
 
+  let(:normal_outcome) do
+    {
+      headings: headings,
+      service_fees: service_fees,
+      output_payments: output_payments,
+      other_fees: other_fees,
+    }
+  end
+  let(:mentor_outcome) do
+    normal_outcome.merge(headings: mentor_focussed_headings)
+  end
+  let(:ect_outcome) do
+    normal_outcome.merge(headings: ect_focussed_headings)
+  end
+
+  def set_precision(hash, rounding)
+    with_rounding(hash) { |sub_hash, k, v| sub_hash[k] = v.round(rounding) }
+  end
+
+  def with_rounding(hash, &blk)
+    hash.each do |k, v|
+      yield(hash, k, v) if v.is_a?(BigDecimal)
+      case hash[k]
+      when Hash
+        with_rounding(hash[k], &blk)
+      when Array
+        # I know this is a hash, if it wasn't this wouldn't work and you'd have to handle the values more elegantly
+        hash[k].each do |service_fee_hash|
+          with_rounding(service_fee_hash, &blk)
+        end
+      else
+        next
+      end
+    end
+  end
```

Comment:
What is this for? It is used to set the precision of our results to be 2 every time, apparently. Should we perhaps change the `normal_outcome` to have the correct precision? Or the orchestrator to spit out numbers with precision 2?

Reply:
It's a recursive descent method to walk through the payment breakdown hash and convert any BigDecimal values to a specific rounding. Only used for this breakdown hash (which is why the comment) and these tests (which is why the location). Since this is financial data, it shouldn't be rounded until we need to display it, but the tests don't need to work to 15 digits of precision.

This is set to round to 2 for now. It's only for testing purposes. The display output will be 2 or 0 digits, depending on requirements. (Options for these are currently being discussed.)